File: examples/ruby/response_action/response_action.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Demonstrates returning canned HTTP responses with MockServer.
#
# Prerequisites: MockServer running on localhost:1080
#   docker run -d -p 1080:1080 mockserver/mockserver

require 'mockserver-client'
require 'net/http'
require 'json'

include MockServer

client = Client.new('localhost', 1080)

# ------------------------------------------------------------------
# 1. Simple response with body only
# ------------------------------------------------------------------
client.when(
  HttpRequest.request(path: '/hello')
).respond(
  HttpResponse.response(body: 'Hello, World!')
)
puts '1. Created expectation: GET /hello -> 200 "Hello, World!"'

# ------------------------------------------------------------------
# 2. Response with status code, headers, and cookies
# ------------------------------------------------------------------
client.when(
  HttpRequest.new(method: 'POST', path: '/login')
).respond(
  HttpResponse.new(
    status_code: 201,
    reason_phrase: 'Created',
    headers: [
      KeyToMultiValue.new(name: 'Content-Type', values: ['application/json']),
      KeyToMultiValue.new(name: 'X-Custom', values: ['example'])
    ],
    cookies: [
      KeyToMultiValue.new(name: 'session', values: ['abc-123'])
    ],
    body: '{"status": "logged_in"}'
  )
)
puts '2. Created expectation: POST /login -> 201 with headers and cookies'

# ------------------------------------------------------------------
# 3. Response with a delay
# ------------------------------------------------------------------
client.when(
  HttpRequest.request(path: '/slow')
).respond(
  HttpResponse.new(
    status_code: 200,
    body: 'delayed response',
    delay: Delay.new(time_unit: 'SECONDS', value: 2)
  )
)
puts '3. Created expectation: GET /slow -> 200 after 2s delay'

# ------------------------------------------------------------------
# 4. Different responses for the same path (first match wins)
# ------------------------------------------------------------------
client.when(
  HttpRequest.request(path: '/counter'),
  times: Times.exactly(1)
).respond(
  HttpResponse.response(body: 'first', status_code: 200)
)

client.when(
  HttpRequest.request(path: '/counter'),
  times: Times.exactly(1)
).respond(
  HttpResponse.response(body: 'second', status_code: 200)
)
puts '4. Created expectations: GET /counter -> "first" then "second"'

# ------------------------------------------------------------------
# 5. Custom status code and reason phrase
# ------------------------------------------------------------------
client.when(
  HttpRequest.new(method: 'POST', path: '/teapot')
).respond(
  HttpResponse.new(status_code: 418, reason_phrase: "I'm a teapot")
)
puts "5. Created expectation: POST /teapot -> 418 I'm a teapot"

# ------------------------------------------------------------------
# Verify by sending a real request
# ------------------------------------------------------------------
puts "\n--- Sending test request to GET /hello ---"
uri = URI('http://localhost:1080/hello')
response = Net::HTTP.get_response(uri)
puts "Status: #{response.code}"
puts "Body:   #{response.body}"

# ------------------------------------------------------------------
# Clean up
# ------------------------------------------------------------------
client.reset
puts "\nAll expectations cleared."

File: examples/python/response_action/response_action.py
#!/usr/bin/env python3
"""MockServer Python client -- response action examples.

Each function creates an expectation that returns a canned HTTP response.
Run this script against a MockServer on localhost:1080.
"""

from mockserver import (
    Body,
    ConnectionOptions,
    Delay,
    Expectation,
    HttpRequest,
    HttpResponse,
    MockServerClient,
    Times,
)


MOCK_HOST = "localhost"
MOCK_PORT = 1080


def response_with_body_only(client: MockServerClient) -> None:
    """Return a plain-text body for every request (no request matcher)."""
    client.when(
        HttpRequest()
    ).respond(
        HttpResponse.response("some_response_body")
    )
    print("expectation created: response with body only")


def response_with_status_code_and_reason(client: MockServerClient) -> None:
    """Return a custom status code and reason phrase."""
    client.when(
        HttpRequest(method="POST", path="/some/path")
    ).respond(
        HttpResponse(status_code=418, reason_phrase="I'm a teapot")
    )
    print("expectation created: 418 I'm a teapot")


def response_with_header(client: MockServerClient) -> None:
    """Return a response with a custom header."""
    client.when(
        HttpRequest(path="/some/path")
    ).respond(
        HttpResponse.response("some_response_body")
        .with_header("Content-Type", "text/plain")
    )
    print("expectation created: response with header")


def response_with_cookie(client: MockServerClient) -> None:
    """Return a response with a Set-Cookie header."""
    client.when(
        HttpRequest(path="/some/path")
    ).respond(
        HttpResponse.response("some_response_body")
        .with_header("Content-Type", "text/plain")
        .with_cookie("Session", "97d43b1e-fe03-4855-926a-f448eddac32f")
    )
    print("expectation created: response with cookie")


def response_with_json_body(client: MockServerClient) -> None:
    """Return a JSON body."""
    client.when(
        HttpRequest(method="GET", path="/api/user")
    ).respond(
        HttpResponse.response()
        .with_status_code(200)
        .with_header("Content-Type", "application/json")
        .with_body(Body.json({"id": 1, "name": "Alice", "active": True}))
    )
    print("expectation created: JSON body response")


def response_with_delay(client: MockServerClient) -> None:
    """Return a response after a 2-second delay."""
    client.when(
        HttpRequest(path="/slow/path")
    ).respond_with_delay(
        HttpResponse.response("delayed_body"),
        Delay(time_unit="SECONDS", value=2),
    )
    print("expectation created: response with 2s delay")


def respond_differently_for_same_request(client: MockServerClient) -> None:
    """First request returns 200, next two return 204."""
    client.when(
        HttpRequest(path="/some/path"),
        times=Times(remaining_times=1, unlimited=False),
    ).respond(
        HttpResponse(status_code=200)
    )
    client.when(
        HttpRequest(path="/some/path"),
        times=Times(remaining_times=2, unlimited=False),
    ).respond(
        HttpResponse(status_code=204)
    )
    print("expectation created: different responses for same path")


def response_with_connection_options(client: MockServerClient) -> None:
    """Return a response that suppresses content-length and connection headers."""
    client.upsert(
        Expectation(
            http_request=HttpRequest(path="/some/path"),
            http_response=HttpResponse(
                body="some_response_body",
                connection_options=ConnectionOptions(
                    suppress_content_length_header=True,
                    suppress_connection_header=True,
                ),
            ),
        )
    )
    print("expectation created: response with connection options")


def main() -> None:
    with MockServerClient(MOCK_HOST, MOCK_PORT) as client:
        client.reset()

        response_with_body_only(client)
        response_with_status_code_and_reason(client)
        response_with_header(client)
        response_with_cookie(client)
        response_with_json_body(client)
        response_with_delay(client)
        respond_differently_for_same_request(client)
        response_with_connection_options(client)

        print("\nAll response-action expectations created successfully.")


if __name__ == "__main__":
    main()
